explorer/interpreter/pattern_analysis.cpp

```diff
@@ -0,0 +1,307 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "explorer/interpreter/pattern_analysis.h"
+
+#include <set>
+
+using llvm::cast;
+using llvm::dyn_cast;
+using llvm::isa;
+
+namespace Carbon {
+
+auto AbstractPattern::kind() const -> Kind {
+  if (auto* pattern = value_.dyn_cast<const Pattern*>()) {
+    return Compound;
+  }
+  if (auto* value = value_.dyn_cast<const Value*>()) {
+    if (isa<TupleValue, AlternativeValue, BoolValue>(value)) {
+      return Compound;
+    }
+    return Primitive;
+  }
+  CARBON_CHECK(value_.is<const WildcardTag*>());
+  return Wildcard;
+}
+
+auto AbstractPattern::discriminator() const -> std::string_view {
+  CARBON_CHECK(kind() == Compound);
+  if (auto* pattern = value_.dyn_cast<const Pattern*>()) {
+    if (auto* alt_pattern = dyn_cast<AlternativePattern>(pattern)) {
+      return alt_pattern->alternative_name();
+    }
+  } else if (auto* value = value_.dyn_cast<const Value*>()) {
+    if (auto* alt = dyn_cast<AlternativeValue>(value)) {
+      return alt->alt_name();
+    } else if (auto* bool_val = dyn_cast<BoolValue>(value)) {
+      return bool_val->value() ? "true" : "false";
+    }
+  }
+  return {};
+}
+
+auto AbstractPattern::NumElements() const -> int {
```

Comment:
FWIW, I don't understand why `kind()` or `discriminator()` accessor naming, but then `NumElements` here (they seem to have similar characteristics). I don't have a strong opinion one way or the other, but it may make sense to change one or the other.

Reply:
Yeah, I find this awkward too. `kind` and `discriminator` are trying to follow the "use lowercase for property names" rule. It's not clear to me whether `NumElements` should follow that rule too or not, but it doesn't feel especially property-like to me, so I used the "normal" CamelCase convention.

Reply:
Renamed to `elements_size` per offline discussion.